crypto/evp/digest.c

```diff
@@ -1017,8 +1015,11 @@ void EVP_MD_free(EVP_MD *md)
 {
     int i;
 
-    if (md == NULL)
+    if (md == NULL || md->origin == ORIG_GLOBAL)
         return;
+    if (md->origin == ORIG_METH) {
+        /* EVP_MD_meth_free just calls this function. */
+    }
```

Comment:
I'd put a comment about EVP_MD_meth_free above the `if (md == NULL ...)` line and remove this no-op `if`

Reply:
Someone needs to decide if `ORIG_METH` methods should be refcounted or not.  *They are currently not*.  I see two options:

1.  Make `EVP_get_digestbyname()` and `EVP_get_cipherbyname()` increment the refcount on `ORIG_METH` methods.
2.  Do not refcount `ORIG_METH` methods, and make it so that this function refuses to free `ORIG_METH` methods.  The consequence, as far as I can see, is that `EVP_MD_meth_free()` and `EVP_CIPHER_meth_free()` should not call
`EVP_MD_free()` and `EVP_CIPHER_free()`, and should instead do the freeing as they do in 1.1.1.

Reply:
I'd go with 2

Reply:
I don't yet have informed feelings about either approach, but master currently calls `EVP_MD_free`, for example and I am not sure what the impact of moving away from that has.  @levitte, you made the change as part of #9758, do you recall why?

Reply:
> @levitte, you made the change as part of #9758, do you recall why?

Because we were going to deprecate the whole set of `EVP_MD_meth_` and `EVP_CIPHER_meth_` functions, and we still needed something to free a fetched `EVP_MD` or `EVP_CIPHER`.  At the time, the semantics for the `_meth_` and non-`_meth_` variants didn't differ.  With the changes made here, however, the semantics *are* different.

Reply:
>  With the changes made here, however, the semantics are different.

yes, thanks.  so i wasn't missing anything.